### PR TITLE
drivers: add gp2y10xx dust sensor

### PR DESCRIPTION
**Pull request by jeandudey**
### Contribution description

This adds a driver for Sharp GP2Y1010AU0F/GP2Y1014AU0F Compact Optical Dust Sensors. There are some magic values in the code that don't make much sense but these are what I found on Waveshare et all Arduino code to convert ADC readings to ug/m3.

### Testing procedure

An example test is provided for the driver, it can be flashed as:

- `make -C tests/driver_gp2y10xx`

Make sure you set the `GP2Y10XX_PARAM_AOUT` (ADC line), `GP2Y10XX_PARAM_ILED`/`ILED_LEVEL` (GPIO pin connected to sensor's ILED) and `GP2Y10XX_PARAM_VREF` (ADC voltage reference) parameters for correct readings.

You may need to give it some smoke to see readings though since the sensor can't detect particulate below 50 ug/m3 levels, that said, it isn't recommended to use it as a fire/smoke detector because it's an unreliable device.

<sup>:warning: I do not endorse smoking a cigarette (as you may create an addiction) just to test it but a friend of a friend tested it and works :eyes:, melting a solder wire containing a flux core can be detected too though as an alternative</sup>

### CODE REVIEW

File: tests/driver_gp2y10xx/main.c [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    while (1) {
+        res = gp2y10xx_read_density(&dev, &density);
+        if (res == GP2Y10XX_OK) {
+            printf("Dust density %"PRIi16" ug/m3\n", density);
+        }
+        else {
+            printf("Error reading data. err: %d\n", res);
+        }
+
+        xtimer_usleep(SLEEP_USEC);
```
Comment (benpicco) [inline tests/driver_gp2y10xx/main.c:57]:
```suggestion
        xtimer_msleep(1);
```

Comment (benpicco) [reply tests/driver_gp2y10xx/main.c:57]: actually that will generate a lot of output - maybe rather use 250 ms here

**Review by benpicco — COMMENTED** [review]
looks good, appears to be a pretty simple device.

File: drivers/gp2y10xx/gp2y10xx.c [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            assert(0);
+            break;
+    }
+
+    return (1 << exp);
+}
+
+/* calculates the ADC sample value by filtering it with all
+ * previous ADC readings */
+static int32_t _adc_filter(gp2y10xx_t *dev, int32_t adc_line_value) {
```
[the remark this answers is not in the document]
Comment (jeandudey) [reply drivers/gp2y10xx/gp2y10xx.c:88]: Changed to filter on `gp2y10xx_read_density` multiple times, it improved the sensor output and the code.

**Comment by jeandudey** [comment]
Comments addressed.

Tested the code on an esp32-wrover-kit, however `adc_set_attenuation` (esp32 specific, to read 3.3v on ADC) needs to be set by the test application, would be better if that bit of code were on an `#ifdef CPU_ESP32` block?

**Comment by benpicco** [comment]
Looks like you didn't push all the changes yet :wink: 

> `adc_set_attenuation` (esp32 specific, to read 3.3v on ADC) needs to be set by the test application,

How does that work?
According to the [doc](https://doc.riot-os.org/adc__arch_8h.html#a6786cc4a1c5602bf2466638ab3e8dff6) `ADC_ATTENUATION_11_DB` should be the default and give the full range.

But yea, the ADC API needs a rework (#13247 :eyes: )

**Comment by jeandudey** [comment]
Some test output from last changes:

```
2020-11-16 10:35:15,059 # Dust density 0 ug/m3
2020-11-16 10:35:15,315 # Dust density 85 ug/m3
2020-11-16 10:35:15,571 # Dust density 222 ug/m3
2020-11-16 10:35:15,827 # Dust density 81 ug/m3
2020-11-16 10:35:16,083 # Dust density 81 ug/m3
2020-11-16 10:35:16,323 # Dust density 92 ug/m3
2020-11-16 10:35:16,579 # Dust density 79 ug/m3
2020-11-16 10:35:16,835 # Dust density 79 ug/m3
2020-11-16 10:35:17,091 # Dust density 83 ug/m3
2020-11-16 10:35:17,347 # Dust density 83 ug/m3
```

**Comment by jeandudey** [comment]
>According to the doc ADC_ATTENUATION_11_DB should be the default and give the full range.

Oops, didn't read it very well I was resettting it to the default :P

File: drivers/gp2y10xx/gp2y10xx.c [hunk 3]
```diff
@@ -144,17 +104,21 @@ int gp2y10xx_read_density(gp2y10xx_t *dev, uint16_t *density)
...
           dev->params.iled_pin);
 
-    /* turn ILED on/off and wait a little bit to read the ADC */
-    _iled_on(dev);
-    xtimer_usleep(ILED_PULSE_WAIT_US);
-    if ((adc_value = adc_sample(dev->params.aout, dev->params.adc_res)) < 0) {
+    for (unsigned i = 0; i < CONFIG_GP2Y10XX_MAX_READINGS; i++) {
+        /* turn ILED on/off and wait a little bit to read the ADC */
+        _iled_on(dev);
+        xtimer_usleep(ILED_PULSE_WAIT_US);
```
Comment (benpicco) [inline drivers/gp2y10xx/gp2y10xx.c:110]:
```suggestion
        xtimer_usleep(ILED_PULSE_WAIT_US);
        _iled_on(dev);
```

The way I understand the current ADC API, the samping should only start with `adc_sample()` - so you could turn off the LED while you wait for the next period.

Comment (jeandudey) [reply drivers/gp2y10xx/gp2y10xx.c:110]: The sensor takes it's time in between the led gets powered on and detects particulate, so it's at the end of the 280 us that it generates data

Comment (benpicco) [reply drivers/gp2y10xx/gp2y10xx.c:110]: Then you can also move the `_iled_on(dev)` / `_iled_off(dev)` out of the loop

Comment (jeandudey) [reply drivers/gp2y10xx/gp2y10xx.c:110]: Changin the iled_on/off out of the loop changes the output a little bit:

Current one:

```
2020-11-16 11:11:33,532 # Dust density 90 ug/m3
2020-11-16 11:11:33,771 # Dust density 85 ug/m3
2020-11-16 11:11:34,028 # Dust density 88 ug/m3
2020-11-16 11:11:34,283 # Dust density 83 ug/m3
2020-11-16 11:11:34,539 # Dust density 81 ug/m3
2020-11-16 11:11:34,796 # Dust density 90 ug/m3
```

Out of the loop:

```
2020-11-16 11:10:47,043 # Dust density 0 ug/m3
2020-11-16 11:10:47,299 # Dust density 424 ug/m3
2020-11-16 11:10:47,539 # Dust density 455 ug/m3
2020-11-16 11:10:47,795 # Dust density 424 ug/m3
2020-11-16 11:10:48,051 # Dust density 440 ug/m3
2020-11-16 11:10:48,291 # Dust density 448 ug/m3
2020-11-16 11:10:48,548 # Dust density 429 ug/m3
2020-11-16 11:10:48,803 # Dust density 435 ug/m3
2020-11-16 11:10:49,043 # Dust density 426 ug/m3
2020-11-16 11:10:49,299 # Dust density 451 ug/m3
```

Comment (benpicco) [reply drivers/gp2y10xx/gp2y10xx.c:110]: Huh that is weird - without debug output there should be almost no time between `_iled_off(dev)` and `_iled_on(dev)`

Comment (jeandudey) [reply drivers/gp2y10xx/gp2y10xx.c:110]: In the [datasheet](https://www.sparkfun.com/datasheets/Sensors/gp2y1010au_e.pdf) they specify the 280 us sampling time, so IMO (while slower) should be left inside the loop.

Screenshot from 2020-11-16 11-16-58

Comment (benpicco) [reply drivers/gp2y10xx/gp2y10xx.c:110]: Ah so it needs an on -> off transition to start sampling.
Then maybe add a short delay after `_iled_off(dev)` to make sure the off time is not too short on fast CPUs.

Comment (jeandudey) [reply drivers/gp2y10xx/gp2y10xx.c:111]: Added a short 20us delay at the `_iled_off` just to be sure :eyes: :

**Review by benpicco — APPROVED** [review]
looks good, please squash

**Comment by benpicco** [comment]
[Travis](https://travis-ci.org/github/RIOT-OS/RIOT/builds/743917927) is not happy

**Comment by jeandudey** [comment]
Addressed Travis comments :eyes:

**Comment by benpicco** [comment]
Murdock is not happy yet either.

**Review by benpicco — COMMENTED** [review]
one more nitpick: defining a type for the level (instead of using `int`) makes the params struct easier to understand.

**Comment by jeandudey** [comment]
Comments addressed, thanks for noticing it!

**Comment by jeandudey** [comment]
Thanks for the review :wink:

**Pull request merged**